### DISCUSSION TITLE
BIRT engine is null in Docker deploy

### DISCUSSION THREAD

**Discussion by eduardogmisiuk** [eclipse-birt/birt#999] [Q&A] [answered]
I am trying to use BIRT in my application that will run in Docker, without success. The application works when running in Intellij IDEA, but when I try to run the application in the Docker container, I got this error:
```
java.lang.NullPointerException: Cannot invoke "org.eclipse.birt.report.engine.api.IReportEngineF
actory.createReportEngine(org.eclipse.birt.report.engine.api.EngineConfig)" because "factory" is null
```

I am using Java JDK 17 with Dropwizard and Maven. My pom is:
```yml
    <dependency>
      <groupId>org.apache.poi</groupId>
      <artifactId>poi-ooxml</artifactId>
      <version>3.17</version>
    </dependency>
    <dependency>
      <groupId>org.apache.poi</groupId>
      <artifactId>poi</artifactId>
      <version>3.17</version>
    </dependency>
    <dependency>
      <groupId>org.eclipse.birt.runtime</groupId>
      <artifactId>org.eclipse.osgi.services</artifactId>
      <version>3.4.0.v20140312-2051</version>
    </dependency>
    <dependency>
      <groupId>org.eclipse.birt.runtime</groupId>
      <artifactId>com.ibm.icu</artifactId>
      <version>52.1.1.v201501240615</version>
    </dependency>
    <dependency>
      <groupId>org.eclipse.birt.runtime</groupId>
      <artifactId>org.eclipse.birt.runtime</artifactId>
      <version>4.4.2</version>
      <exclusions>
        <exclusion>
          <groupId>org.mozilla.javascript</groupId>
          <artifactId>javascript</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.apache.poi</groupId>
          <artifactId>poi-ooxml</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.apache.poi</groupId>
          <artifactId>poi</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.eclipse.birt.runtime</groupId>
          <artifactId>org.eclipse.osgi.services</artifactId>
        </exclusion>
        <exclusion>
          <groupId>org.eclipse.birt.runtime</groupId>
          <artifactId>com.ibm.icu</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
```

The ManagedObject from Dropwizard that initialized BIRT:
```java
  public void start() throws Exception {
    try {
      EngineConfig config = new EngineConfig();

      Platform.startup(config);

      IReportEngineFactory factory =
          (IReportEngineFactory)
              Platform.createFactoryObject(IReportEngineFactory.EXTENSION_REPORT_ENGINE_FACTORY);
      BirtManagedObject.birtEngine = factory.createReportEngine(config);
      BirtManagedObject.birtEngine.changeLogLevel(Level.WARNING);
    } catch (Exception e) {
      e.printStackTrace();
      throw e;
    }
  }
```

**Comment by wimjongman** [comment]
Are you running inside OSGi? [1]


[1] https://github.com/eclipse/birt/blob/e488c851ab2bf2fcdce0f5e3723c5a2a4b4ca324/core/org.eclipse.birt.core/src/org/eclipse/birt/core/framework/Platform.java#L247

**Reply by eduardogmisiuk** [reply]
I think I'm not, since I don't initialize anything for BIRT besides what I posted. Is it something I should start in Java or another container?

**Comment by dm-ion** [comment]
I know you said it works from IntelliJ, but I would probably recommend updating to the latest BIRT - 4.9.

It looks like you are using BIRT 4.4.2. This version may not work as intended on JDK 17 at all due to reflective changes in JDK 16, have you done a jdeps scan to check for potential internal access issues?

All that said, not sure what your issue could be, we run BIRT inside a docker (non OSGI) on JDK 17 without issues. If I had to guess I would say you are probably missing some dependencies from your project. (eg: org.eclipse.osgi.services which you've excluded)

Our BIRT engine instantiation code looks similar to yours.

```
@Override
  public IReportEngine getObject() throws Exception {
    // Create the birt temp dirs in case they do not exist already (as per the web.xml file)
    final File birtFolder = new File(birtReportDirectory);
    if (!birtFolder.exists()) {
      FileUtils.forceMkdir(birtFolder);
    }

    final EngineConfig engineConfig = new EngineConfig();
    engineConfig.setLogger(LOGGER);

    // Required due to a bug in BIRT
    RegistryProviderFactory.releaseDefault();
    Platform.startup(engineConfig);

    final IReportEngineFactory reportEngineFactory = (IReportEngineFactory) Platform.createFactoryObject(IReportEngineFactory.EXTENSION_REPORT_ENGINE_FACTORY);
    final IReportEngine newBirtEngine = reportEngineFactory.createReportEngine(engineConfig);

    // try to call dummy report to init birt report engine
    try (InputStream dummyReport = CharSource.wrap(BirtReportConstants.BIRT_REPORT_DUMMY).asByteSource(StandardCharsets.UTF_8).openBufferedStream()) {
      newBirtEngine.openReportDesign(dummyReport);
    }

    birtReportEngine = newBirtEngine;
    return newBirtEngine;
  }
```

**Reply by eduardogmisiuk** [reply]
I tried to use the 4.9.0 version, but did not succeed. There is also an open issue about the Maven build for this version that is broken in Maven Central. Are you using Maven? If so, could you share how you made it work?

**Reply by dm-ion** [reply]
As you mentioned it's not on maven central, I built it myself.

**Reply by dm-ion** [reply]
Did you try adding back in the osgi dependency that you excluded? org.eclipse.osgi.services

**Reply by eduardogmisiuk** [reply]
The first attempt was without excluding dependencies, but it didn't work, then I tried to exclude and add newer versions, which did not work too. I am trying to build BIRT, not really successful by now.

**Comment by wimjongman** [comment, marked as answer]
You cannot build BIRT from maven ATM. We have nothing recent on maven. 

You can download the BIRT Standalone Runtime:

https://download.eclipse.org/birt/downloads/drops/latest/birt-runtime-4.9.0-20220315.zip

* unzip it
* go the commandline in the ReportEngine directory
* run: genReport.bat samples\hello_world.rptdesign

This will create a report in the samples directory.

You will have to drop your drivers and other custom stuff in the ReportEngine libs directory

Then drop your report design in the samples directory and test if they work

**Reply by wimjongman** [reply]
@eduardogmisiuk any updates?

**Reply by eduardogmisiuk** [reply]
Saddly we decided to drop BIRT for now from the application. In the future we might use it again, since the PDF generation can get prioritized again.

**Reply by wimjongman** [reply]
Ok. Thanks for the feedback.
